### ISSUE TITLE
Persistence and shared subscriptions

### ISSUE THREAD

**Issue by Llorx** [erlio/vernemq#618]
- VerneMQ Version: 1.3.0
- OS: Server Debian - Client Windows

Status:
 - 2 Clients with shared subscriptions and persistent session subscribed to the same topic.
 - 1 Client publishing to the topic.

At first you see how the messages are received randomly between both clients. Everything is fine. When you close one client, messages are still sent to closed clients like they were active, so only half of the messages are processed actually.

As per [this message](https://github.com/erlio/vernemq/issues/230#issuecomment-274321440), closed clients should not receive messages and only be sent to active clients.

When you disable persistence, everything works as expected.

Small test that reproduces the problem with a default VerneMQ installation.
```
let mqtt = require('mqtt');

messages = 10;
test_topic = 'strangeTopic';
let client1 = {
    url: 'mqtt://10.2.3.31:1883',
    options: { clean: false, clientId: 'test_client1' },
}
let client2 = {
    url: 'mqtt://10.2.3.31:1883',
    options: { clean: false, clientId: 'test_client2' },
}

let publisher = {
    url: 'mqtt://10.2.3.31:1883',
    options: {}
}

let conn1 = mqtt.connect(client1.url, client1.options);
let conn2 = null;

conn1.on('connect', (err, res) => {
    conn1.subscribe(`$share/test/${test_topic}`, (err, res) => {
        conn2 = mqtt.connect(client2.url, client2.options);
        conn2.on('connect', (err, res) => {
            conn2.subscribe(`$share/test/${test_topic}`, (err, res) => {
                conn1.end();
                publish();
            });
            conn2.on('message', (topic, payload, packet) => {
                console.log(`message received on client2 with topic: ${topic} and payload: ${payload}`)
            })
        });
    });
    conn1.on('message', (topic, payload, packet) => {
        console.log(`message received on client1 with topic: ${topic} and payload: ${payload}`)
    });
});

let publish = () => {
    let conn3 = mqtt.connect(publisher.url);
    conn3.on('connect', (err, res) => {
        // a message per second
        for (let i = 0; i < messages; i++) {
            setTimeout(() => {
                conn3.publish(test_topic, 'test payload', (err, res) => {
                    console.log(`published ${i + 1} messages`);
                 });
            }, (i + 1) * 1000);
        }
    })
}
```

**Comment by ioolkos** [comment]
@Llorx thanks! ... trying to find out what the original intention was... :)
@larshesel any insights?

**Comment by Llorx** [comment]
@ioolkos There was no intention. We are 2 users testing our client-server against the same broker. I suddenly noticed that my client was not delivering messages correctly to the servers, so we started to see what happened and the reason was that my coworker closed his server.

**Comment by Llorx** [comment]
The real intention is to have a cluster of servers receiving messages in a shared subscription from the broker, and some times we will need to close or open servers depending on the amount of messages processed. We want persistent connections to not lose messages if all servers lose connection to the broker, but we want only "active" servers to receive messages from the broker.

**Comment by larshesel** [comment]
Hi @Llorx I haven't had time to look into this yet, but just to understand are the messages actually *lost* or are they ending up in the offline queue for the client that is no longer connected?

**Comment by Llorx** [comment]
Hi @larshesel. I guess the second, but I have to check it. Will reply with more info.

**Comment by Llorx** [comment]
Hi, @larshesel. My teammate has done some checks. It goes to the offline queue as when the server is reconnected to the broker, it receives the queued messages.

**Comment by larshesel** [comment]
Hi @Llorx , thanks for the checking that. I also confirmed that what you experience is a bug which prevents the delivery to the online queues. I'll prepare a PR with the fix for this today.

**Comment by Llorx** [comment]
Yeah, I've taken a look to the fix and saw the argument flip xD

We will use the nightly. Thank you.

**Comment by larshesel** [comment]
The fix #620 was just merged to master and will therefore be included in the next nightly build.
